### PR TITLE
Add Multinomial Naive Bayes to linfa-bayes

### PR DESCRIPTION
**Pull request by sgrigory**
Closes "Multinomial Naive Bayes" part of #183

- Refactors `linfa-bayes` crate so that one can add different kinds of Naive Bayes without much code overlap
- Adds Multinomial Naive Bayes, in addition to Gaussian Naive Bayes which was already there. Implementation is inspired by `sklearn.naive_bayes.MultinomialNB` and it's used to compute the expected results for test cases

### Note on the type system

Initially I was planning to move the functionality which depends on the kind of Naive Bayes into a trait (let's call it `TraitNB` here), as suggested in #183, and implement this trait differently for Multinomial  and Gaussian NB. Then I would implement common NB functionality for any type T which implements `TraitNB`. However, I ran into a problem related to [specialization](https://github.com/rust-lang/rust/issues/31844): it seems impossible to implement traits like `Fit` for a generic type T, since it already has a generic implementation in [src/param_guard.rs](https://github.com/rust-ml/linfa/blob/b73a0fd9743a0df49baa251ba7707c36a964b693/src/param_guard.rs#L55). Please correct me if I'm missing some simple workaround here.

I therefore chose a different pattern: 
- Base types are enums - `base_nb::BaseNb` and `hyperparams::NbModel`
- Common functionality (e.g. `BaseNb::predict_inplace` method) is implemented for the whole enum
- Specific functionality is implemented for structures which correspond to the variants of enum, e.g. `GaussianNb::joint_log_likelihood` and `MultinomialNb::joint_log_likelihood`

I'm not sure this is the optimal solution and would be happy to hear your feedback on this.

### Note on dataset for the examples

[winequality](https://github.com/rust-ml/linfa/blob/master/datasets/data/winequality-red.csv.gz) doesn't seem to be the best dataset to illustrate the usage of Multinomial NB and it gives quite low accuracy in the example. Perhaps we could create a separate issue to add to `linfa-datasets` a dataset on which application of Multinomial NB would make sense, e.g. something like [20 Newsgroups data set](http://qwone.com/~jason/20Newsgroups/)?

### CODE REVIEW

File: algorithms/linfa-bayes/README.md [hunk 1]
```diff
@@ -10,11 +10,12 @@
 
 `linfa-bayes` currently provides an implementation of the following methods: 
 
-- Gaussian Naive Bayes (GaussianNB)
+- Gaussian Naive Bayes ([`GaussianNb`](crate::GaussianNb))
+- Multinomial Naive Nayes ([`MultinomialNb`](crate::GaussianNb))
```
Comment (YuhanLiin) [inline algorithms/linfa-bayes/README.md:14]: The link should say `MultinomialNb`

File: algorithms/linfa-bayes/src/base_nb.rs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    // Returns a subset of x corresponding to the class specified by `ycondition`
+    pub fn filter(x: ArrayView2<F>, y: ArrayView1<L>, ycondition: &L) -> Array2<F> {
+        // We identify the row numbers corresponding to the class we are interested in
+        let index = y
+            .into_iter()
+            .enumerate()
+            .filter_map(|(i, y)| match *ycondition == *y {
+                true => Some(i),
+                false => None,
+            })
```
Comment (YuhanLiin) [inline algorithms/linfa-bayes/src/base_nb.rs:129]: You can do `(*y_condition == *y).some().map(|_| i)`

**Comment by YuhanLiin** [comment]
You can still use a Bayes trait to give you implementations for `predict_inplace` and `fit`. You just can't blanket impl for your trait, so you need to have individual impls of `PredictInplace` and `Fit` that act as thin wrappers. The trait can look like this:
```
trait NaiveBayes: FitWith { // FitWith impls have no common code, so they can be implemented for each individual type
  fn joint_log_likelihood() -> T;

  fn predict_inplace() {
    // call joint_log_likelihood
  }

  fn fit() {
    // call fit_with
  }
}
```
Then you impl `Fit` and `PredictInplace` for each NB type and just wrap the `NaiveBayes` methods with the same name, which minimizes amount of duplicated code. `filter` can just be a common function. Using a trait prevents needing to use an enum, which IMO adds unnecessary corner cases. This is especially true for the param type, where many methods panic when called with the wrong variant. The bonus is that the trait doesn't need to be public, since we don't blanket impl on top of it.

**Comment by sgrigory** [comment]
@YuhanLiin Thanks for the detailed reply! That makes sense, let me try to refactor the code as you suggested and update this PR

**Comment by bytesnake** [comment]
thank you for working on this

> Then I would implement common NB functionality for any type T which implements TraitNB. 

you can also introduce a trait `Distribution` (for loglikelihood estimation) and implement that for `univariate::{Gaussian, Multinomial}` structs containing the necessary statistics. The `NaiveBayes` algorithm would then be bounded to `Distribution` and contains a vector of that, one for each feature

File: algorithms/linfa-bayes/src/base_nb.rs [hunk 3]
```diff
@@ -178,24 +46,50 @@ where
...
+}
+
+// Returns a subset of x corresponding to the class specified by `ycondition`
+pub fn filter<F: Float, L: Label + Ord>(
+    x: ArrayView2<F>,
+    y: ArrayView1<L>,
+    ycondition: &L,
+) -> Array2<F> {
+    // We identify the row numbers corresponding to the class we are interested in
+
```
Comment (YuhanLiin) [inline algorithms/linfa-bayes/src/base_nb.rs:80]: Remove space

File: algorithms/linfa-bayes/src/base_nb.rs [hunk 4]
```diff
@@ -3,152 +3,20 @@ use ndarray_stats::QuantileExt;
...
-#[derive(Debug, Clone)]
-pub enum BaseNb<F, L> {
-    MultinomialNb(MultinomialNb<F, L>),
-    GaussianNb(GaussianNb<F, L>),
-}
-
-
-impl<F, L, D, T> Fit<ArrayBase<D, Ix2>, T, NaiveBayesError> for NbValidParams<F, L>
+// Common functionality of fitted Naive Bayes models
+pub trait NaiveBayes<'a, F, L, D>
```
Comment (YuhanLiin) [inline algorithms/linfa-bayes/src/base_nb.rs:11]: I don't think this trait and `NaiveBayesValidParams` need to be public.

**Comment by YuhanLiin** [comment]
Can you move the examples on the param types to the main NB types? I'd rather have detailed docs on the main NB types than the param types. It's also a good replacement for the ignore examples we currently have for the NB types.

File: algorithms/linfa-bayes/src/base_nb.rs [hunk 5]
```diff
@@ -48,8 +46,17 @@ where
...
 
+// Trait computing log-likelihood for Naive Bayes models
+pub(crate) trait NaiveBayesLogLikelihood<F, L>
+where
+    F: Float,
+    L: Label + Ord,
+{
+    fn joint_log_likelihood(&self, x: ArrayView2<F>) -> HashMap<&L, Array1<F>>;
+}
+
```
Comment (YuhanLiin) [inline algorithms/linfa-bayes/src/base_nb.rs:57]: No need to have a separate trait for log likelihood.

Comment (sgrigory) [reply algorithms/linfa-bayes/src/base_nb.rs]: @YuhanLiin Sure, I've moved log-likelihood back into the `NaiveBayes<'a, F, L, D>` trait. The only side effect is that now one needs to explicitly specify type parameter D when calling `joint_log_likelihood`, e.g. here:  https://github.com/rust-ml/linfa/pull/190/files#diff-20110e51d7148056af774f667a88af96f6abf6f9f3d3e79961e2098b6f11f441R261-R262. Compiler can't automatically infer type `D`  because `joint_log_likelihood` doesn't depend on it, but the whole `NaiveBayes` trait does.

Comment (YuhanLiin) [reply algorithms/linfa-bayes/src/base_nb.rs]: In `NaiveBayes` can you move type parameter `D` from the trait to the `predict_inplace` method, since it's only relevant for that method? That should fix the type inference issues.

Comment (sgrigory) [reply algorithms/linfa-bayes/src/base_nb.rs]: Yep, that worked, thanks

**Review by YuhanLiin — APPROVED** [review]

**Comment by YuhanLiin** [comment]
Thanks for the PR!

**Comment by sgrigory** [comment]
> Thanks for the PR!

Thanks for the guidance!
